tangelo/linq/target/target_qulacs.py

```diff
@@ -74,36 +77,85 @@ def simulate_circuit(self, source_circuit: Circuit, return_statevector=False, in
+                    python_statevector = state.get_vector()
+                    if self._noise_model:
+                        n_success += 1
+                        sample = state.sampling(1)[0]
+                        samples[sample] = samples.get(sample, 0) + 1
+                    else:
+                        self._current_state = python_statevector
+                        if self.n_shots is not None:
+                            samples = Counter(state.sampling(self.n_shots))
+                        else:
+                            frequencies = self._statevector_to_frequencies(python_statevector)
+                            self.all_frequencies = dict()
+                            for meas, val in frequencies.items():
+                                self.all_frequencies[measure + meas] = val
+
                 if initial_statevector is not None:
                     state.load(initial_statevector)
                 else:
                     state.set_zero_state()
+                n_attempts += 1
+
+            if self.n_shots is not None:
+                self.all_frequencies = {measure + self._int_to_binstr(k, source_circuit.width): v / self.n_shots
+                                        for k, v in samples.items()}
+
+            if n_attempts == max_attempts:
+                raise ValueError(f"desired_meas_result was not measured after {n_attempts} attempts")
+
+            return (self.all_frequencies, python_statevector) if return_statevector else (self.all_frequencies, None)
 
-        # To save mid-circuit measurement results
         elif save_mid_circuit_meas:
             n_meas = source_circuit.counts.get("MEASURE", 0)
             samples = dict()
             for _ in range(self.n_shots):
                 translated_circuit.update_quantum_state(state)
-                measurement = "".join([str(state.get_classical_value(i)) for i in range(n_meas)])
+                measure = "".join([str(state.get_classical_value(i)) for i in range(n_meas)])
                 sample = self._int_to_binstr(state.sampling(1)[0], source_circuit.width)
-                bitstr = measurement + sample
+                bitstr = measure + sample
                 samples[bitstr] = samples.get(bitstr, 0) + 1
+
+                # Will only be true if self.n_shots=1
+                if return_statevector:
+                    python_statevector = state.get_vector()
+                    self._current_state = python_statevector
                 if initial_statevector is not None:
                     state.load(initial_statevector)
                 else:
                     state.set_zero_state()
+
             self.all_frequencies = {k: v / self.n_shots for k, v in samples.items()}
             return (self.all_frequencies, python_statevector) if return_statevector else (self.all_frequencies, None)
 
-        # All other cases for shot-based simulation
+        # If saving mid-circuit measurements is not necessary or a noise model
```

Comment:
I strongly suggest you start with the easy cases in your file, or what users are most likely to use.

What I personally dislike most is when I'm looking for something simple and the software makes me read about a bunch of complicated things that were not necessary. This is an open-source projects so many people may read or try to modify this code.